### ISSUE TITLE
cpp-tests job is flaky: python pulsar_test.py fails with Segmentation fault

### ISSUE THREAD

**Issue by lhotari** [apache/pulsar#14714]
The cpp-tests job is flaky. 

```
./run-unit-tests.sh: line 99:  2746 Segmentation fault      (core dumped) python pulsar_test.py
```

Example failure:
https://github.com/apache/pulsar/runs/5566228154?check_suite_focus=true#step:11:1313

Logs: https://github.com/apache/pulsar/suites/5678205919/artifacts/186533781
https://transfer.sh/DUTouQ/test-logs.zip

**Comment by lhotari** [comment]
@BewareMyPower Do you have a chance to take a look at the cpp-tests failures? There's currently a lot of flakiness.

**Comment by lhotari** [comment]
the above segfault isn't the only issue, there's also flaky tests, recent example: https://github.com/apache/pulsar/runs/5567332966?check_suite_focus=true#step:11:568

```
FAILED TESTS (4/284):
    1405 ms: ./main ClientTest.testConnectTimeout (try #1)
    1323 ms: ./main ClientTest.testConnectTimeout (try #2)
    2666 ms: ./main ProducerTest.testSendAsyncCloseAsyncConcurrentlyWithLazyProducers (try #1)
     225 ms: ./main ReaderTest.testReferenceLeak (try #14148
```

**Comment by BewareMyPower** [comment]
Issues of other cpp flaky tests:
- https://github.com/apache/pulsar/issues/13849
- https://github.com/apache/pulsar/issues/14665
- https://github.com/apache/pulsar/issues/14719

I'm going to work on these cpp tests first.

**Comment by BewareMyPower** [comment]
This issue will be closed once https://github.com/apache/pulsar/pull/14823 is merged. But I'm not very sure the flaky Python tests are all related with the issue. You can reopen this issue if Python tests are still flaky.

**Comment by BewareMyPower** [comment]
Reopen this issue for https://github.com/apache/pulsar/runs/5687970436?check_suite_focus=true

**Comment by BewareMyPower** [comment]
[Another test failure](https://github.com/apache/pulsar/runs/5854461931?check_suite_focus=true)

```
2022-04-06 16:36:04.579 INFO  [140456268633856] MultiTopicsConsumerImpl:95 | Successfully Subscribed to Topics
2022-04-06 16:36:08.081 INFO  [140457878607616] ClientImpl:496 | Closing Pulsar client with 3 producers and 1 consumers
2022-04-06 16:36:08.081 INFO  [140457878607616] ProducerImpl:661 | [persistent://public/default/my-python-pattern-consumer-1-partition-0, standalone-0-463] Closing producer for topic persistent://public/default/my-python-pattern-consumer-1-partition-0
2022-04-06 16:36:08.082 INFO  [140457878607616] ProducerImpl:704 | [persistent://public/default/my-python-pattern-consumer-1-partition-0, standalone-0-463] Closed producer
2022-04-06 16:36:08.082 INFO  [140457878607616] ProducerImpl:661 | [persistent://public/default/my-python-pattern-consumer-1-partition-1, standalone-0-464] Closing producer for topic persistent://public/default/my-python-pattern-consumer-1-partition-1
2022-04-06 16:36:08.083 INFO  [140457878607616] ProducerImpl:704 | [persistent://public/default/my-python-pattern-consumer-1-partition-1, standalone-0-464] Closed producer
2022-04-06 16:36:08.083 INFO  [140457878607616] ProducerImpl:661 | [persistent://public/default/my-python-pattern-consumer-2-partition-0, standalone-0-465] Closing producer for topic persistent://public/default/my-python-pattern-consumer-2-partition-0
2022-04-06 16:36:08.084 INFO  [140457878607616] ProducerImpl:704 | [persistent://public/default/my-python-pattern-consumer-2-partition-0, standalone-0-465] Closed producer
2022-04-06 16:36:08.084 INFO  [140457878607616] ProducerImpl:661 | [persistent://public/default/my-python-pattern-consumer-2-partition-1, standalone-0-466] Closing producer for topic persistent://public/default/my-python-pattern-consumer-2-partition-1
2022-04-06 16:36:08.085 INFO  [140457878607616] ProducerImpl:704 | [persistent://public/default/my-python-pattern-consumer-2-partition-1, standalone-0-466] Closed producer
2022-04-06 16:36:08.085 INFO  [140457878607616] ProducerImpl:661 | [persistent://public/default/my-python-pattern-consumer-2-partition-2, standalone-0-467] Closing producer for topic persistent://public/default/my-python-pattern-consumer-2-partition-2
2022-04-06 16:36:08.085 INFO  [140457878607616] ProducerImpl:704 | [persistent://public/default/my-python-pattern-consumer-2-partition-2, standalone-0-467] Closed producer
2022-04-06 16:36:08.085 INFO  [140457878607616] ProducerImpl:661 | [persistent://public/default/my-python-pattern-consumer-3-partition-0, standalone-0-468] Closing producer for topic persistent://public/default/my-python-pattern-consumer-3-partition-0
2022-04-06 16:36:08.086 INFO  [140457878607616] ProducerImpl:704 | [persistent://public/default/my-python-pattern-consumer-3-partition-0, standalone-0-468] Closed producer
2022-04-06 16:36:08.086 INFO  [140457878607616] ProducerImpl:661 | [persistent://public/default/my-python-pattern-consumer-3-partition-1, standalone-0-469] Closing producer for topic persistent://public/default/my-python-pattern-consumer-3-partition-1
2022-04-06 16:36:08.087 INFO  [140457878607616] ProducerImpl:704 | [persistent://public/default/my-python-pattern-consumer-3-partition-1, standalone-0-469] Closed producer
2022-04-06 16:36:08.087 INFO  [140457878607616] ProducerImpl:661 | [persistent://public/default/my-python-pattern-consumer-3-partition-2, standalone-0-470] Closing producer for topic persistent://public/default/my-python-pattern-consumer-3-partition-2
2022-04-06 16:36:08.087 INFO  [140457878607616] ProducerImpl:704 | [persistent://public/default/my-python-pattern-consumer-3-partition-2, standalone-0-470] Closed producer
2022-04-06 16:36:08.088 INFO  [140457878607616] ProducerImpl:661 | [persistent://public/default/my-python-pattern-consumer-3-partition-3, standalone-0-471] Closing producer for topic persistent://public/default/my-python-pattern-consumer-3-partition-3
2022-04-06 16:36:08.088 INFO  [140457878607616] ProducerImpl:704 | [persistent://public/default/my-python-pattern-consumer-3-partition-3, standalone-0-471] Closed producer
2022-04-06 16:36:08.088 INFO  [140457878607616] ConsumerImpl:999 | [persistent://public/default/my-python-pattern-consumer-3-partition-3, my-pattern-consumer-sub, 8] Closing consumer for topic persistent://public/default/my-python-pattern-consumer-3-partition-3
2022-04-06 16:36:08.090 INFO  [140457878607616] ConsumerImpl:1055 | [persistent://public/default/my-python-pattern-consumer-3-partition-3, my-pattern-consumer-sub, 8] Closed consumer 8
..................................................../run-unit-tests.sh: line 99:  2772 Segmentation fault      (core dumped) python pulsar_test.py
```

Attach these logs to locate where it crashed.

**Comment by BewareMyPower** [comment]
Another test failure

```
2022-06-08 15:00:43.627 INFO  [140029770565440] ProducerImpl:665 | [persistent://public/default-2/my-python-topics-consumer-3-partition-2, standalone-0-560] Closing producer for topic persistent://public/default-2/my-python-topics-consumer-3-partition-2
2022-06-08 15:00:43.627 INFO  [140029770565440] ProducerImpl:708 | [persistent://public/default-2/my-python-topics-consumer-3-partition-2, standalone-0-560] Closed producer
2022-06-08 15:00:43.628 INFO  [140029770565440] ProducerImpl:665 | [persistent://public/default-2/my-python-topics-consumer-3-partition-3, standalone-0-562] Closing producer for topic persistent://public/default-2/my-python-topics-consumer-3-partition-3
2022-06-08 15:00:43.629 INFO  [140029770565440] ProducerImpl:708 | [persistent://public/default-2/my-python-topics-consumer-3-partition-3, standalone-0-562] Closed producer
2022-06-08 15:00:43.629 INFO  [140029770565440] ConsumerImpl:1005 | [persistent://public/default-2/my-python-topics-consumer-3-partition-2, my-topics-consumer-sub, 7] Closing consumer for topic persistent://public/default-2/my-python-topics-consumer-3-partition-2
2022-06-08 15:00:43.630 INFO  [140029770565440] ConsumerImpl:1062 | [persistent://public/default-2/my-python-topics-consumer-3-partition-2, my-topics-consumer-sub, 7] Closed consumer 7
..................................................../run-unit-tests.sh: line 93:  2879 Segmentation fault      (core dumped) python3 pulsar_test.py
```

It looks like a problem related to the multi topic consumer.

**Comment by BewareMyPower** [comment]
@Jaudouard Could you open another issue for it? I think it's not related to the flaky Python tests. `unsubscribe` is an API to delete the subscription of the consumer. It does not close the connection so the file descriptors are still open, while `close` just closes all related connections.

**Comment by Jaudouard** [comment]
Thanks for your answer, we'll make our case in a dedicated issue today.

Our problem might be that when calling `unsubscribe()`, it is not possible to call `close() `afterwards to properly close the connection.

And if we skip the `unsubscribe()`, pulsar topics accumulate in a dangerous way

**Comment by BewareMyPower** [comment]
Yeah, I have seen similar feedback from other people. I'll confirm if this behavior is right soon.

> And if we skip the unsubscribe(), pulsar topics accumulate in a dangerous way

Could you explain it in detail? It's possible to delete the topic even if it has a subscription unless the subscription is active.

**Comment by Jaudouard** [comment]
> Could you explain it in detail? It's possible to delete the topic even if it has a subscription unless the subscription is active.

It is indeed possible but it requires a pulsar-admin call doesn't it ? Hence with auth and access to this command, which is not trivial in our case

**Comment by BewareMyPower** [comment]
Got it. I think one workaround is to modify the `brokerDeleteInactiveTopicsMode` config of the broker and acknowledge messages.

BTW, I just tested `close` after `unsubscribe` with the Java client, it's okay to do that. I'll look into the issue with the C++ client soon.

**Comment by Jaudouard** [comment]
Thanks for your help, the workaround might well be helpful if we manage to apply it only to our product topics.

I've created a simple bug report: https://github.com/apache/pulsar/issues/16006
